### PR TITLE
[NPU] Support DSpark compact verify graph and folded verify epilogue

### PR DESCRIPTION
**Pull request by 2044145178**
## Motivation

DSpark on the Ascend DSV4 backend currently falls back to a uniform verify window or eager-side post-processing for parts of speculative decoding. This leaves two sources of avoidable overhead:

1. the target model verifies the full draft window even when the SPS policy selects a shorter per-request verify length;
2. accept/finalize and hidden/KV commit remain outside the target verify graph.

This PR enables the compact DSpark path on NPU, adds token-tiered graph replay for ragged target verify, and folds the eligible verify epilogue into the graph.

### 1. DSpark compact verify on NPU

- Build packed/ragged verify windows from per-request verify lengths selected by the DSpark planner.
- Add NPU-safe compact row mapping, verify-ID gathering, cache-location handling, and compact-to-strided scatter.
- Construct DSV4 SWA, sparse-attention, C4/C128 compressor, position, state, and page-table metadata from the live compact layout.
- Preserve eager fallback when the selected attention backend does not support ragged verify graphs.

### 2. Compact target-verify graph

- Advertise ragged verify graph support for the DSV4 Ascend backend.
- Select target-verify graphs by compact token tier while retaining fixed graph-owned buffers.
- Refresh query indptr, sequence lengths, cache/page tables, compressor locations/state, and sparse indices before replay.
- Add an NPU-local bucketed-layout cache so the Ascend and DSV4 metadata consumers reuse the same derived layout.
- Use the safe last-location Triton path for DSV4 NPU allocation metadata.

### 3. Verify epilogue graph

- Add a DSV4 NPU-specific target-verify tail hook that uses the committed/live prefix lengths from DSV4 metadata.
- Fold greedy accept/finalize and eligible hidden/KV commit into the compact target-verify graph.
- Keep sampling, grammar-constrained decoding, logits adjustments, and other ineligible cases on the existing eager epilogue path.
- Keep the adaptation NPU/DSV4-specific so the existing CUDA behavior remains unchanged.

## Accuracy Tests

- GPQA-Diamond accuracy: **0.91** with DSpark on Ascend NPU.
- The compact path preserves the target-model verification result; it changes the number of draft tokens verified and where graph-safe post-processing executes.

## Speed Tests and Profiling

No standalone performance number is included in this PR. The expected gain comes from reducing target-verify tokens according to the SPS-selected compact lengths and removing eligible verify epilogue launches from the host path.

## Validation

- Python syntax compilation passed for all Python files changed by this PR.
- Added scheduler coverage for exact eager token counts when the backend does not support ragged verify graphs.

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy).
- [ ] Provide speed benchmark results according to [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

---
### CI States

Latest PR Test (Base): :x: [Run #33484164263](https://github.com/sgl-project/sglang/actions/runs/33484164263)
Latest PR Test (Extra): :x: [Run #33484164236](https://github.com/sgl-project/sglang/actions/runs/33484164236)
Latest PR Test (AMD ROCm 7.2): :x: [Run #33484164360](https://github.com/sgl-project/sglang/actions/runs/33484164360)

### CODE REVIEW

File: python/sglang/srt/hardware_backend/npu/attention/ascend_backend.py [hunk 1]
```diff
@@ -756,7 +773,26 @@ def _apply_cuda_graph_metadata(
...
+                if ragged_layout.bs != bs:
+                    ragged_layout = get_npu_bucketed_ragged_verify_layout(
+                        spec_info=spec_info,
+                        layout=ragged_layout,
+                        padded_bs=bs,
+                        cap=int(
+                            getattr(
+                                spec_info,
+                                "draft_token_num",
+                                self.speculative_num_draft_tokens,
```
Comment (2044145178) [inline python/sglang/srt/hardware_backend/npu/attention/ascend_backend.py]: redundant code

File: python/sglang/srt/hardware_backend/npu/attention/ascend_dsv4_backend.py [hunk 2]
```diff
@@ -99,6 +103,30 @@ def _build_explicit_state_block_table(
...
 
+    @staticmethod
+    def _ragged_verify_layout(forward_batch: ForwardBatch):
+        if not forward_batch.forward_mode.is_target_verify():
+            return None
+        layout = resolve_ragged_verify_layout(forward_batch)
+        if layout is None:
+            return None
+        padded_bs = int(forward_batch.batch_size)
+        if padded_bs < layout.bs:
```
Comment (2044145178) [inline python/sglang/srt/hardware_backend/npu/attention/ascend_dsv4_backend.py]: over-defensive

File: python/sglang/srt/models/deepseek_v4_dspark.py [hunk 3]
```diff
@@ -712,10 +713,10 @@ def __init__(
         use_multi_stream = (
             envs.SGLANG_OPT_USE_MULTI_STREAM_OVERLAP.get()
```
Comment (randgun) [inline python/sglang/srt/models/deepseek_v4_dspark.py]: or envs.SGLANG_NPU_USE_MULTI_STREAM.get()

File: python/sglang/srt/hardware_backend/npu/extra_ops_loader.py [hunk 4]
```diff
@@ -25,7 +25,36 @@ class OpLibSpec:
 
 
 class TorchOpLoader:
-    """Load a standalone .so into ``torch.ops`` and validate its operators."""
+    """
+    Loader for PyTorch custom operators from shared libraries.
+
```
Comment (randgun) [inline python/sglang/srt/hardware_backend/npu/extra_ops_loader.py:30]: revert

File: python/sglang/srt/speculative/dspark_components/dspark_draft_sampler.py [hunk 5]
```diff
@@ -202,7 +202,7 @@ def _eager(reason):
     )
```
Comment (randgun) [inline python/sglang/srt/speculative/dspark_components/dspark_draft_sampler.py:202]: Delete dlash_disaggregation.py

File: python/sglang/srt/hardware_backend/npu/attention/ascend_dsv4_backend.py [hunk 6]
```diff
@@ -1104,6 +1158,21 @@ def _build_dsv4_graph_replay_ctx(self, forward_batch: ForwardBatch):
             if is_idle_replay:
                 live_seq_lens_cpu = torch.zeros_like(raw_seq_lens_cpu)
                 final_seq_lens_cpu = live_seq_lens_cpu
+            elif ragged_layout is not None:
```
Comment (randgun) [inline python/sglang/srt/hardware_backend/npu/attention/ascend_dsv4_backend.py]: ragged should be a sub branch of dspark

File: python/sglang/srt/hardware_backend/npu/attention/ascend_dsv4_backend.py [hunk 7]
```diff
@@ -780,6 +818,7 @@ def forward_c4_indexer(
 class DeepseekV4AscendAttnBackend(
     AscendAttnBackend, C4IndexerAscendBackendMixin, CompressorAscendBackendMixin
 ):
+    supports_ragged_verify_graph = True
```
Comment (randgun) [inline python/sglang/srt/hardware_backend/npu/attention/ascend_dsv4_backend.py]: delete
